### ISSUE TITLE
Open/focus a note from command line / desktop URL handler?

### ISSUE THREAD

**Issue by bkidwell** [TriliumNext/trilium#649]
I would like some mechanism to:

- Retrieve a URL or ID from the current note displayed in Trilium
- Paste that URL or ID somewhere outside Trilium (for example, in a more advanced task manager)
- Customize my desktop or other apps so that the note ID can be sent back into a running Trilium instance and open/focus the target note ID

For example:

1. On my desktop, a URL scheme is registered: "trilium:$id" calls Trilium with $id as a command line parameter.
2. [In Trilium] Copy current note URL to clipboard.
3. [In a separate task manager] Paste URL.
4. Executing note URL (on my desktop) tells causes the target note to be opened in Trilium.

Is there any way to setup something like this?

**Comment by zadam** [comment]
Hello, custom URL scheme looks like a good way to interact with the Trilium.

Electron seems to [support this](https://electronjs.org/docs/api/protocol) so it doesn't even look that difficult. I will do some experimentation with this ...

**Comment by meichthys** [comment]
[Trilium has entered maintenance mode](https://github.com/zadam/trilium/issues/4620#issuecomment-2094495698). Future enhancements will be addressed in TrilumNext: https://github.com/TriliumNext/trilium/issues/4988

**Comment by meichthys** [comment]
I think this is the relevant electron documentation: https://www.electronjs.org/docs/latest/api/protocol

This feature would be really helpful in a business environment where other  documentation needs to reference pages in trilium. For example, an ERP system should be able to link directly to specific Trilium documentation page. This is possible with a Trilium web URL, but on desktop a Trilium:// protocol handler should be registered so the page can be opened in the trilium desktop app by default instead of a web browser.

https://github.com/TriliumNext/Trilium/issues/5728 would be related to this since we would need a way to copy a note's path using the Trillium:// protocol.

**Comment by Vertex3x** [comment]
I tried a lot of things, like scripting of calling window - search (wmctrl + xdotool), getting note content from server API / sqlite directly using python wrapper, but this is not a solution.
I need this feature on linux desktop (trillium:// urls support), to call any notes. 
Ready to sponsor it. Can we do this please?

**Comment by eliandoran** [comment]
@Vertex3x ,

Please specify exactly your needs (URLs to support plus expected behavior). You can already assign a bounty to it using Issuehunt, if you wish.

**Comment by Vertex3x** [comment]
@eliandoran , thanks for reply
Added bounty to this issue.
Request description:

**Feature Specification: Deep Linking / Custom URI Scheme (trilium://) for Desktop**

**Overview & Use Case**
I am looking to integrate Trilium more tightly with my OS environment and external programs. I need the ability to generate a universal link to a specific note, paste that link into an external application, and have my OS open that exact note inside the Trilium desktop app when the link is clicked.

My primary environment is Linux desktop, so robust support for Linux is a strict requirement for this bounty, though ideally, the solution should be cross-platform (Electron supports this natively).

**Functional Requirements**
**Link Generation (UI):**
1. Provide a mechanism within the Trilium UI (e.g., right-clicking a note in the tree, or a button in the note header) to "Copy App Link" to the clipboard.
2.  The copied link should follow a standard URI scheme format, such as trilium://note/[$noteId].

**OS Protocol Registration:**
The Trilium desktop application must register itself as the default handler for the trilium:// protocol scheme at the OS level.

**In-App Protocol Handling & Navigation:**
1. Already Running (Single Instance): If Trilium is already running in the background or minimized, clicking the trilium:// link should bring the existing Trilium window to the foreground/focus and immediately open the target $noteId.
2. (OPTIONAL) Cold Start: If Trilium is completely closed, clicking the link should launch the application, initialize the workspace, and navigate directly to the target $noteId.

**Acceptance Criteria for Bounty Payout:**
1. Executing xdg-open trilium://note/<id> (or clicking the link in a Linux terminal/browser/task manager) successfully opens the note.
2. The feature works seamlessly on Linux desktop environments without requiring the user to manually write custom bash scripts to parse the URI.
3. The app handles the link correctly when it is already running (when it is launched from a closed state - optional).

**Comment by meichthys** [comment]
It's got to be me, but I've tried to donate to this issue a few times and keep getting this error:
![Screenshot_20260322-170120.png](https://github.com/user-attachments/assets/d27139b7-20be-484e-845f-ce884933e18a)

Maybe it's just a grapheneOS 
 or Firefox problem. Not sure, but looking forward to having this implemented!

**Comment by eliandoran** [comment]
@meichthys , I suspect it's a problem with IssueHunt, but unfortunately I don't have a solution for it. I even tried reaching out to them and nothing. I didn't find a suitable alternative either.

**Comment by BeatLink** [comment]
Hey guys, mind if i take a crack at it?

**Comment by eliandoran** [comment]
@BeatLink , go ahead.

**Comment by thebasedcapital** [comment]
I am working on this. Plan:
- register the desktop app as a `trilium://` protocol handler
- parse `trilium://note/<noteId>` on cold start, macOS `open-url`, and Electron `second-instance` flows
- focus the existing window and navigate to the requested note when the app is already running
- add a note-tree context menu action to copy `trilium://note/<noteId>` app links

**Comment by thebasedcapital** [comment]
Opened PR: https://github.com/TriliumNext/Trilium/pull/9688

**Comment by DavideZanonArt** [comment]
Submitted a focused desktop-side implementation in #9712.

It registers the `trilium://` protocol, handles `trilium://note/<id>` on startup / second-instance / macOS `open-url`, and reuses the existing `openInSameTab` path to focus the requested note.

**Comment by vayungodara** [comment]
Submitted #9713 with the full feature set from @Vertex3x's acceptance criteria:

- `trilium://note/<noteId>` protocol handler — cold start on Windows/Linux (`process.argv`), macOS `open-url` event, `second-instance` for the already-running case, and `--new-window` deep links
- "Copy app link to clipboard" context menu item under the existing "Copy note path" entry (Electron-only, hidden on root)
- 24 unit tests for the URL parser covering canonical/triple-slash forms, percent-encoding round trips, path-traversal/XSS/shell-meta rejection, bounded length, mixed-case preservation
- Smoke-tested on macOS via `pnpm dev`; CI green; gemini-code-assist review addressed
- Docs section added under `Desktop Installation.md`

AI-assisted disclosure: drafted with Claude Code following the repo's `CLAUDE.md` (pure-function parser extracted with unit tests per the guidance there). All design choices and edge cases reviewed before submission. Happy to iterate on feedback.

**Comment by eliandoran** [comment]
I've decided to unilaterally refund all the bounties that we had on IssueHunt.

When I first set up the bounties, the concept made perfect sense since people could easily "vote with their wallets" their favorite features which would also be an incentive mostly to the dev team to implement said features. Now with the AI bubble, the barrier to creating PRs is almost zero which means that we have had a significant influx of contributions meant only to receive said bounty. This is not something we encourage because it puts an unequal burden on us as reviewers, since the cost of a review is higher than the development cost (when AI is used).

When refunding I got this message from IssueHunt saying that the Stripe fee will not be refunded. Apologies for the lost funds, but perhaps it's for the best. I would prefer to encourage a tip culture instead; when a feature lands that you are happy with, simply use the normal donation channels. This also unblocks your funds since sometimes it can take a lot of time to implement a feature.

If you have any issues or concerns, let me know.

**Comment by meichthys** [comment]
@eliandoran Couldn't we just limit pull requests to contributors? This is how Frappe Framework does it. If someone wants to become a contributor, they can fork, make changes and link to the changes in an issue. If the changes are reasonable and concise, then users could be added as contributors(?).

I suppose it's too late now, but at least this way the main Trilium contributors would have the ability to claim the bounties - which are a good indicator of feature demand.

**Comment by eliandoran** [comment]
@meichthys , 

I plan to make two important changes to the workflow that are similar:

* Create a concrete policy so that we don't get flooded with PRs, from bounties or otherwise (the number of PRs that we have to review has grown significantly since LLMs became mainstream). That policy would involve working on assigned issues only.
* Feature requests will be moved out of Issues and into GitHub Discussions / Ideas, where they can be voted for visibility. Out of 724 issues, 452 are feature requests/wishes; some of them might never materialize so we need to distinguish them somehow (by visibility). Accepted feature requests can be transferred back as issues so that they can be picked up by someone.

From what I've seen the bounty itself attracts too much unwanted attention, even if we try to restrict it. Regardless, I've reached the conclusion that it's better for people to donate after the fact.

**Comment by meichthys** [comment]
Ok, seems reasonable to me. Thanks for explaining.
